typeset/derive/config.satyg

```diff
@@ -1,14 +1,104 @@
 @import: line
 
-type label-position =
-    | LabelLeft
-    | LabelRight
+type label-pos-t =
+| LabelLeft
+| LabelRight
 
-type derive-config = (|
-    default-label-position : label-position;
-    default-line : derivation-line;
+type derive-config-t = (|
+    default-label-position : label-pos-t;
+    default-line : DeriveLine.u;
     bar-top-bottom-margin : length;
     label-left-right-margin : length;
     bar-thickness : length;
     content-left-right-padding : length;
 |)
+
+module DeriveConfig : sig
+
+    val default : derive-config-t
+
+    val with-default-label-position : label-pos-t -> derive-config-t -> derive-config-t
+    val with-default-line : DeriveLine.u -> derive-config-t -> derive-config-t
+    val with-bar-top-bottom-margin : length -> derive-config-t -> derive-config-t
+    val with-label-left-right-margin : length -> derive-config-t -> derive-config-t
+    val with-bar-thickness : length -> derive-config-t -> derive-config-t
+    val with-content-left-right-padding : length -> derive-config-t -> derive-config-t
+
+end = struct
+
+    type t = (|
+        default-label-position : label-pos-t;
+        default-line : DeriveLine.u;
+        bar-top-bottom-margin : length;
+        label-left-right-margin : length;
+        bar-thickness : length;
+        content-left-right-padding : length;
+    |)
+
+
+    let default = (|
+        default-label-position = LabelLeft;
+        default-line = DeriveLine.standard-line;
+        bar-top-bottom-margin = 2pt;
+        label-left-right-margin = 3pt;
+        bar-thickness = 0.5pt;
+        content-left-right-padding = 3pt;
+    |)
+
+    let make config = config
+
+    let with-default-label-position value ast = (|
```

Comment:
Perhaps you can use record update syntax https://github.com/gfngfn/SATySFi/commit/850ba23a4da0271033181496498d8473be2b7445

Reply:
Wow, good know that. Thanks👍

Reply:
Well, do we need those `with-*` functions given that we already have record update syntax?

Reply:
I think it's worth because `with-*` functions are composable.

A better design is possibly define derive-config-t as an opaque type `DeriveConfig.t` and add getter methods like `DeriveConfig.get-default-label-position`. This way we can add more fields to DeriveConfig.t later without changing the interface.

Reply:
Sounds good to me.